staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -649,6 +648,12 @@ func (r *Request) request(fn func(*http.Request, *http.Response)) error {
 		if err != nil {
 			return err
 		}
+		if r.timeout > 0 {
+			if r.ctx == nil {
+				r.ctx = context.Background()
+			}
+			r.ctx, _ = context.WithTimeout(r.ctx, r.timeout)
```

Comment:
this looks like it's applying the timeout per attempt... does this mean a timeout of 10 seconds could get repeated 10 times, resulting in 100 seconds before this returned? I'm not sure that's what the "Change to a timeout based approach." TODO above intended.

Reply:
ah, WithTimeout -> WithDeadline, which doesn't lengthen the deadline, so that's ok.

I don't think it's correct to discard the cancelfunc though:

> Canceling this context releases resources associated with it, so code should call cancel as soon as the operations running in this Context complete

Reply:
> cancel must be called.

Hmm... I had some doubts about that, but looking at it again it does make sense to have it there.